core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -822,25 +822,42 @@ object Cause extends Serializable {
     final case class Filter[E](p: Cause[E] => Boolean) extends Folder[Any, E, Cause[E]] {
       def empty(context: Any): Cause[E] = Cause.empty
 
-      def failCase(context: Any, error: E, stackTrace: StackTrace): Cause[E] = Cause.Fail(error, stackTrace)
+      def failCase(context: Any, error: E, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Fail(error, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
-      def dieCase(context: Any, t: Throwable, stackTrace: StackTrace): Cause[E] = Cause.Die(t, stackTrace)
+      def dieCase(context: Any, t: Throwable, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Die(t, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
-      def interruptCase(context: Any, fiberId: FiberId, stackTrace: StackTrace): Cause[E] =
-        Cause.Interrupt(fiberId, stackTrace)
+      def interruptCase(context: Any, fiberId: FiberId, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Interrupt(fiberId, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
       def bothCase(context: Any, left: Cause[E], right: Cause[E]): Cause[E] =
-        if (p(left)) {
-          if (p(right)) Cause.Both(left, right)
+        if (left.nonEmpty) {
+          if (right.nonEmpty) Cause.Both(left, right)
           else left
-        } else if (p(right)) right
+        } else if (right.nonEmpty) right
```

Comment:
`nonEmpty` is not necessarily cheap, and it's called at least once, potentially twice per method invocation. Should we store it in vals at the start of the method for both left and right?

Reply:
agreed, though not sure this is a real concern.
anyways, it can be optimized by using some kind of short circuiting traversal over the `Cause` or alternatively by caching the result of `isEmpty` (which actually do use a short circuiting traversal)

Reply:
@kyri-petrou , actually in this case we can accumulate the `nonEmptiy` property through the traversal.
the idea is to modify `Cause.Filter` to be a `Folder` producing a `(Cause[E], Boolean)` (instead of just `Cause[E]`),
I'm afraid this requires changing the signature of a public class though.

Reply:
another option which greatly simplifies things **without** modifying any public signature, the filter fold can be modified to guarantee that 'filtered out' parts of the tree are replaced by `Cause.Empty`, this way the `nonEmpty` call can be replaced with `eq Cause.Empty`.
in fact the only thing preventing this from happening is the `Stackless` case, I'm not sure what's the meaning of `Stackless(Empty)` but I suspect it's equivalent to plain `Empty`.

seems to me [zio.Cause#unified](https://github.com/zio/zio/blob/e4469873546995e9351cd72d03f7e734d729c261/core/shared/src/main/scala/zio/Cause.scala#L692-L694) is the only piece of code affected by `stackless` and as far as I can read the code this function returns an empty list both for `Stackless` and `Empty`.

Reply:
> `nonEmpty` is not necessarily cheap, and it's called at least once, potentially twice per method invocation. Should we store it in vals at the start of the method for both left and right?

it's never invoked more than once on either objects, so no real gain pre calculating them

Reply:
Ah yeah my bad you're right